### PR TITLE
Show me the scatters!

### PR DESCRIPTION
**Pull request by cjllanwarne**
- [x] Depends on https://github.com/broadinstitute/wdl4s/pull/186

Shows cool scatter graphs like this!

WDL:
```
task foo {
  Int i
  command { ... }
  output {
    String str_out = i
    Int int_out = i + 1
  }
}

workflow scatter_test {
  Int x = 5
  Int y = 6
  scatter (s in [x, y]) {
    call foo { input: i = s }
  }
}
```

WomGraph
test

### CODE REVIEW

**Comment by mcovarr** [comment]
questions: 

* Why is the `s` in the blue box of type `Any`?
* Shouldn't the `s` in the blue box connect to `i` in the white box?

File: src/main/scala/wdltool/graph/WomGraph.scala [hunk 1]
```diff
@@ -39,42 +41,102 @@ class WomGraph(graphName: String, graph: Graph) {
...
+    inputPort <- graphNode.inputPorts
+    upstreamPort = inputPort.upstream
+  } yield s"${upstreamPort.graphId} -> ${inputPort.graphId}"
+
+  private def portLine(p: GraphNodePort) = s"${p.graphId} [shape=${p.graphShape} label=${p.graphName}];"
+
+  private def portLines(graphNode: GraphNode): String = graphNode match {
+    case s: ScatterNode =>
+      // Don't include the scatter expression input port here since they're added later in `internalScatterNodesAndLinks`
+      // Round up the gathered output ports so it's obvious they're being gathered..
```
Comment (mcovarr) [inline src/main/scala/wdltool/graph/WomGraph.scala]: .. => .

File: src/main/scala/wdltool/graph/WomGraph.scala [hunk 2]
```diff
@@ -39,42 +41,102 @@ class WomGraph(graphName: String, graph: Graph) {
...
+         |subgraph $scatterVariableClusterName {
+         |  style=filled;
+         |  fillcolor=${scatter.graphFillColor};
+         |  ${scatter.scatterVariableMapping.scatterInstantiatedExpression.inputPorts map portLine mkString "\n  "}
+         |  $scatterVariableExpressionId [shape=plaintext label="${scatter.scatterVariableMapping.graphInputNode.name} in ..."]
+         |}
+         """.stripMargin
+
+      val scatterVariableInputLink = s"$scatterVariableExpressionId -> ${scatter.scatterVariableMapping.graphInputNode.singleOutputPort.graphId} [style=dashed ltail=$scatterVariableClusterName arrowhead=empty]"
+      val outputLinks = scatter.outputMapping map { case outputPort @ ScatterGathererPort(_, _, outputToGather, _) =>
```
Comment (mcovarr) [inline src/main/scala/wdltool/graph/WomGraph.scala]: probably just capturing would suffice, the pattern match seems unnecessary.  might even fit on one line.  😄

File: src/main/scala/wdltool/graph/WomGraph.scala [hunk 3]
```diff
@@ -39,42 +41,102 @@ class WomGraph(graphName: String, graph: Graph) {
...
+    val clusterName = nextCluster
 
     val singleNode =
       s"""
          |subgraph $clusterName {
          |  style=filled;
          |  fillcolor=${graphNode.graphFillColor};
-         |  ${portNodes.mkString(sep="\n  ")}
+         |  ${portLines(graphNode)}
+         |  ${internals.nodes.mkString(System.lineSeparator + "  ")}
```
Comment (mcovarr) [inline src/main/scala/wdltool/graph/WomGraph.scala]: why are you `lineSeparator`ing here and `"\n"`ing elsewhere?

File: src/main/scala/wdltool/graph/WomGraph.scala [hunk 4]
```diff
@@ -39,42 +41,102 @@ class WomGraph(graphName: String, graph: Graph) {
     graph.nodes foldMap nodesAndLinks
   }
 
-  private def nodesAndLinks(graphNode: GraphNode): NodesAndLinks = graphNode match {
-    case other => defaultNodesAndLinks(graphNode)
+  private def upstreamLinks(graphNode: GraphNode) = for {
```
Comment (mcovarr) [inline src/main/scala/wdltool/graph/WomGraph.scala]: I liked having the return type

**Comment by cjllanwarne** [comment]
@mcovarr 
- That was a TODO in WDL4S. I fixed that with the [fold-style](https://github.com/broadinstitute/wdl4s/pull/186) WOM graph building.
  - EDIT: Updated the image.
- It's because it's an expression based input node, and the expression's input is "s".

File: src/main/scala/wdltool/graph/package.scala [hunk 5]
```diff
@@ -15,22 +15,19 @@ package object graph {
...
+      case gon: GraphOutputNode => s"${gon.womType.toWdlString} ${gon.name}"
       case expr: ExpressionNode =>
         val inputNames = expr.instantiatedExpression.expression.inputs.mkString(", ")
         s"${expr.womType.toWdlString} ${expr.name} = f($inputNames)"
       case other =>
         throw new Exception(s"womgraph can't find a graphName for GraphNodes of type: ${other.getClass.getSimpleName}")
     })
 
-    def graphId: String = dotSafe("NODE" + graphNode.hashCode().toString)
+    def graphId: String = dotSafe("NODE" + System.identityHashCode(graphNode).toString)
```
Comment (kshakir) [inline src/main/scala/wdltool/graph/package.scala]: Can you refactor this into a `def graphNodeUniqueId: Int = System.identityHashCode(graphNode)`, with a comment that this **should** be good-enough, but on some JVMs could cause unwelcome behavior. https://stackoverflow.com/questions/10645494/can-i-assume-two-objects-with-the-same-system-identityhashcode-are-the-same#10645567

File: src/main/scala/wdltool/graph/package.scala [hunk 6]
```diff
@@ -40,6 +37,6 @@ package object graph {
     }
 
     def graphName: String = dotSafe(graphNodePort.name)
-    def graphId: String = dotSafe("PORT" + graphNodePort.hashCode().toString)
+    def graphId: String = dotSafe("PORT" + System.identityHashCode(graphNodePort).toString)
```
Comment (kshakir) [inline src/main/scala/wdltool/graph/package.scala]: Also, `.toString` isn't necessary.

**Review by kshakir — APPROVED** [review]

File: src/main/scala/wdltool/graph/WomGraph.scala [hunk 7]
```diff
@@ -1,32 +1,38 @@
...
 import java.util.concurrent.atomic.AtomicInteger
 
+import better.files.File
 import cats.data.Validated.{Invalid, Valid}
 import cats.derived.monoid._
 import cats.derived.monoid.legacy._
 import cats.implicits._
 import lenthall.validation.ErrorOr.ErrorOr
-import wdl4s.cwl.{CommandLineTool, CwlCodecs, Workflow}
+import wdl4s.cwl.CwlDecoder.decodeAllCwl
```
Comment (mcovarr) [inline src/main/scala/wdltool/graph/WomGraph.scala]: why importing the method and not the object?

File: src/main/scala/wdltool/graph/package.scala (hunk 5 above)
Comment (mcovarr) [reply src/main/scala/wdltool/graph/package.scala]: 😦

**Review by mcovarr — APPROVED** [review]

**Comment by mcovarr** [comment]
:+1:

**Pull request merged**